File: vigenere/vigenere.py
#!/usr/bin/env python

# Seuil d'IC pour cryptanalyse
# La valeur a prendre depend de la langue naturelle.
# Ici on est en anglais, 0.066 est une bonne valeur pour l'IC.
# https://fr.wikipedia.org/wiki/Indice_de_co%C3%AFncidence
ICseuil = 0.064


# caractere -> numero
def char2num(c):
    return ord(c)-ord('a')

# numero -> caractere
def num2char(i):
    return chr(i+ord('a'))

# teste si c est un lettre a-z
def estLettre(c):
    return (ord('a')<=ord(c) and ord(c)<=ord('z'))


# Chiffre un message
def chiffre(K,M):
    C = []
    ik = 0
    for i in range(len(M)):
        if estLettre(M[i]):
            C.append(num2char((char2num(M[i])+char2num(K[ik]))%26))
            ik = (ik+1)%len(K)
        else:
            C.append(M[i])
    return ''.join(C)


# Inverse une clef
def inverseClef(K):
    return ''.join([num2char((26-char2num(k))%26) for k in K])


# Dechiffre un message
def dechiffre(K,M):
    return chiffre(inverseClef(K),M)


# Calcule les indices de coincidence des k sous-textes
# dans le tableau IC et les lettres majoritaires dans ces
# sous-textes dans le tableau Lmaj
def IC(k,M):
    ik = 0
    cpt = [[0 for _ in range(26)] for _ in range(k)]
    Lmaj = [0 for _ in range(k)]
    for c in M:
        if estLettre(c):
            n = char2num(c)
            cpt[ik][n] += 1
            if cpt[ik][n]>cpt[ik][Lmaj[ik]]:
                Lmaj[ik] = n
            ik = (ik+1)%k
    ICs = [0. for _ in range(k)]
    for i in range(k):
        N = 0
        for j in range(26):
            ICs[i] += cpt[i][j]*(cpt[i][j]-1)
            N += cpt[i][j]
        ICs[i] /= N*(N-1)
    return ICs,Lmaj


# Cryptanalyse un message
def cryptanalyse(M):
    k = 0
    ICmoy = 0
    while ICmoy<ICseuil:
        k += 1
        ICs,Lmaj = IC(k,M)
        ICmoy = sum(ICs)/len(ICs)
    return ''.join([num2char((Lmaj[i]+22)%26) for i in range(k)])

File: vigenere/Vigenere.java
import java.lang.Math;

class Vigenere {
    /* seuil d'IC pour cryptanalyse
       la valeur a prendre depend de la langue naturelle
       ici on est en anglais, 0.066 est une bonne valeur pour l'IC
       https://fr.wikipedia.org/wiki/Indice_de_co%C3%AFncidence */
    private float ICseuil = (float)0.064;
    
    
    /* numero -> caractere */
    private char num2char(int i) {
	return (char)(i+(int)'a');
    }

    /* caractere -> numero */
    private int char2num(char c) {
	return (int)c-(int)'a';
    }
    
    /* teste si c est une lettre a-z */
    private boolean estLettre(char c) {
	return ((int)'a'<=(int)c && (int)c<=(int)'z');
    }


    /* chiffre un message (en place) */
    public void chiffre(char[] clef, char[] message) {
	int ik = 0;
	for (int i=0; i<message.length; ++i)
	    if (estLettre(message[i])) {
		message[i] = num2char((char2num(message[i])+char2num(clef[ik]))%26);
		ik = (ik+1)%clef.length;
	    }
    }


    /* inverse une clef */
    public char[] inverseClef(char[] clef) {
	char[] inv_clef = new char[clef.length];
	for (int i=0; i<inv_clef.length; ++i)
	    inv_clef[i] = num2char((26-char2num(clef[i]))%26);
	return inv_clef;
    }


    /* dechiffre un message (en place) */
    public void dechiffre(char[] clef, char[] message) {
	chiffre(inverseClef(clef),message);
    }


    /* calcule les indices de coincidence des k sous-textes
       dans le tableau IC et les lettres majoritaires dans ces
       sous-textes dans le tableau Lmaj */
    private void IC(int k, char[] message, float[] IC, int[] Lmaj) {
	int[][] cpt = new int[k][26];
	int ik = 0;
	for (int i=0; i<message.length; ++i)
	    if (estLettre(message[i])) {
		int n = char2num(message[i]);
		++cpt[ik][n];
		if (cpt[ik][n]>cpt[ik][Lmaj[ik]])
		    Lmaj[ik] = n;
		ik = (ik+1)%k;
	    }
	for (int i=0; i<k; ++i) {
	    int N = 0;
	    for (int j=0; j<26; ++j) {
		IC[i] += cpt[i][j]*(cpt[i][j]-1);
		N += cpt[i][j];
	    }
	    IC[i] /= N*(N-1);
	}
    }


    /* cryptanalyse un message et renvoie la clef decouverte */
    public char[] cryptanalyse(char[] message) {
	int k = 0;
	float ICmoy = 0;
	float[] IC = new float[1];
	int[] Lmaj = new int[1];
	while (ICmoy<ICseuil) {
	    ++k;
	    IC = new float[k];
	    Lmaj = new int[k];
	    IC(k,message,IC,Lmaj);
	    ICmoy = 0;
	    for (int i=0; i<k; ++i)
		ICmoy += IC[i];
	    ICmoy /= k;
	}
	char[] clef = new char[k];
	for (int i=0; i<k; ++i)
	    clef[i] = num2char((Lmaj[i]+22)%26);
	// +22 : décalage par rapport au 'e'
	return clef;
    }
}
